ping/ping.c

```diff
@@ -555,6 +555,23 @@ main(int argc, char **argv)
 	int target_ai_family = hints.ai_family;
 	hints.ai_family = AF_UNSPEC;
 
+	/*
+	 * Allow to use -I<ifname|scope-id> for IPv6 link-local scope address
+	 * when passing it to getaddrinfo():
+	 * i.e. -I<ifname|scope-id> => %<ifname|scope-id>
+	 */
+	unsigned char buf[sizeof(struct in6_addr)];
+	if (!strchr(target, '%') && inet_pton(AF_INET6, target, buf) > 0 &&
+		(IN6_IS_ADDR_LINKLOCAL(buf) || IN6_IS_ADDR_MC_LINKLOCAL(buf))) {
+		if (rts.device) {
+			strcat(target, "%");
+			strcat(target, rts.device);
```

Comment:
As @metan-ucw reported using `strcat()` is wrong because  of previously `target = argv[argc - 1]`. There is no space to append anything to it. Although it "works" I probably rewrote some env variables passed to the process by kernel or something. I should allocate the buffer, e.g. use `asprintf()` which prints into a newly allocated buffer, but that would require to make `_GNU_SOURCE` mandatory. I wonder if it's ok (we have meson check for it).

Reply:
I think `asprintf()` should be available on all the libc implementations we're likely to be linking against (glibc, MUSL, uClibc, anyway).  However, I think `asprintf()` would be wasteful due to the amount of memory copying it would need to do. We could instead do something like the following (untested) and only allocate memory once:

```C
		char *target;
		if (rts.device) {
			target = (char*)malloc(strlen(rts.device) + strlen(argv[argc - 1]) + 2);
			sprintf(target, "%s%%%s", argv[argc - 1], rts.device);
		} else {
			target = (char*)malloc(strlen(argv[argc - 1]) + 1);
			strcat(target, argv[argc - 1]);
		}
```

Not that it matters significantly for the amount of memory we're talking about here.

Reply:
I don't get why [`asprintf()`](https://man7.org/linux/man-pages/man3/asprintf.3.html) should allocate more memory than `malloc()` + `sprintf()`. 

Actually @metan-ucw suggested to use [his `asprintf()` implementation](https://github.com/gfxprim/gfxprim/blob/master/libs/widgets/gp_string.c#L43) (requires also [proper signature](https://github.com/gfxprim/gfxprim/blob/master/include/widgets/gp_string.h#L39)) in case GNU extension is a problem (I don't think so but it's nice to have a backup).

IMHO this would be needed only `if (rts.device)`, when we modify `target`, right? By default we could just reuse `argv[argc - 1]`.

Reply:
Also, we already depend on GNU extensions - building without it results to many failures in `ping` and `arping`. Just to list some of them:
```
../ping/ping.h:319:9: warning: implicit declaration of function ‘sigaction’ [-Wimplicit-function-declaration]
  319 |         sigaction(signo, &sa, NULL);
../ping/ping.c: In function ‘pr_addr’:
../ping/ping.c:1630:19: error: ‘NI_MAXHOST’ undeclared (first use in this function)
 1630 |         char name[NI_MAXHOST] = "";
      |                   ^~~~~~~~~~
../ping/ping.c:1641:9: warning: implicit declaration of function ‘getnameinfo’ [-Wimplicit-function-declaration]
 1641 |         getnameinfo(sa, salen, address, sizeof address, NULL, 0, getnameinfo_flags | NI_NUMERICHOST);

../arping.c:289:9: warning: implicit declaration of function ‘clock_gettime’ [-Wimplicit-function-declaration]
  289 |         clock_gettime(CLOCK_MONOTONIC, &now);
      |         ^~~~~~~~~~~~~
../arping.c:592:36: error: ‘IFF_LOOPBACK’ undeclared (first use in this function); did you mean ‘IN_LOOPBACKNET’?
  592 |         if (ifflags & (IFF_NOARP | IFF_LOOPBACK)) {
      |                                    ^~~~~~~~~~~~
      |                                    IN_LOOPBACKNET
../arping.c: In function ‘event_loop’:
../arping.c:731:9: warning: implicit declaration of function ‘sigemptyset’ [-Wimplicit-function-declaration]
  731 |         sigemptyset(&mask);
      |         ^~~~~~~~~~~
../arping.c:732:9: warning: implicit declaration of function ‘sigaddset’ [-Wimplicit-function-declaration]
  732 |         sigaddset(&mask, SIGINT);
../iputils_common.c:92:25: error: storage size of ‘ts’ isn’t known
   92 |         struct timespec ts;
      |                         ^~
../iputils_common.c:94:9: warning: implicit declaration of function ‘clock_gettime’ [-Wimplicit-function-declaration]
   94 |         clock_gettime(CLOCK_REALTIME, &ts);
      |         ^~~~~~~~~~~~~
../iputils_common.c:94:23: error: ‘CLOCK_REALTIME’ undeclared (first use in this function)
   94 |         clock_gettime(CLOCK_REALTIME, &ts);
```
=> we should depend on GNU anyway s these are old changes (e.g. ad0256f3543a69ff0d44bca838bade6aac04bcbd - released in [s20200821](https://github.com/iputils/iputils/releases/tag/s20200821) - I'll fix this as well).

Reply:
> I don't get why [asprintf()](https://man7.org/linux/man-pages/man3/asprintf.3.html) should allocate more memory than malloc() + sprintf().

It'll allocate memory each time you call it, and as your code was written you were going to end up allocating memory twice: once for the copy of the target from `argv`, and again to create a new, larger, buffer to hold the target with `%IFNAME` appended.

But really the amount of memory isn't critical here. I was more trying to avoid using `_GNU_SOURCE`. But I think it's supported by all the libc implementations we care about. (glibc and MUSL support it, as does uClibc afaik. Are there others?)

Reply:
> I was more trying to avoid using `_GNU_SOURCE`

I'm sorry I didn't wait for you before merge https://github.com/iputils/iputils/pull/402, but we have really depended on `_GNU_SOURCE` for quite long time

> ut I think it's supported by all the libc implementations we care about. (glibc and MUSL support it, as does uClibc afaik. Are there others?)

others: [dietlibc](https://www.fefe.de/dietlibc/) looks to be dead. [μClinux](https://en.wikipedia.org/wiki/%CE%9CClinux) as well.

Android moved some time ago to use iputils s20121221 (and they didn't upgrade), thus we can ignore it, but it also supports asprintf()
https://android.googlesource.com/platform/bionic.git/+/refs/heads/android-t-preview-1/libc/stdio/stdio.cpp#736
https://android.googlesource.com/platform/external/ping/+/refs/heads/master
https://android.googlesource.com/platform/external/iputils/+/refs/heads/android-t-preview-1/README.version